File: Sales_Billing/src/dplyr/_create_sales_view.R
library(dplyr)
library(readr)
library(janitor)
library(here)


# Path
path <- here("Sales_Billing")


# Functions
filter_last_date <- function(df) {
  last_date <- last(df$date) # find the last date
  df <- df |>
    filter(df$date == last_date)
  return(df)
}


change_format <- function(df) {
  df <- df |>
    mutate(
      plant = "0180",
      sales_org = "5668",
      acc_assig_grp = case_when(
        Type == "FERT" ~ "01",
        Type == "HAWA" ~ "02",
        Type == "HALB" ~ "03",
        Type == "ROH" ~ "04",
      ),
      material_group1 = case_when(
        Customer == "HMC" ~ "183",
        Customer == "KMC" ~ "191",
        Customer == "KIA" ~ "191",
        Customer == "GM" ~ "105",
        Customer == "RSM" ~ "193",
        Customer == "Ssangyong" ~ "180",
        Customer == "SYM" ~ "180",
        Customer == "Others" ~ "199",
        Customer == "MOBIS" ~ "317",
        Customer == "Mobis" ~ "317",
        Customer == "MANDO" ~ "384",
        Customer == "Mando" ~ "384",
        Customer == "Webasto" ~ "376",
        Customer == "VT" ~ "360",
        Customer == "Conti" ~ "394",
      ),
      material_group2 = "",
      material_group3 = "",
    )
  return(df)
}


main <- function() {
  # Filenames
  input_file <- here(path, "data", "Sales View Creation Request.csv")
  output_1 <- here("C", "LSMW", "Sales_view.txt")
  output_2 <- here(path, "output", "Sales_view.txt")

  # Read data
  df <- read_csv(input_file, show_col_types = FALSE) |>
    clean_names()
  df <- read_csv("C:/Users/uiv09452/OneDrive - Vitesco Technologies/Desktop/VBA Extract/Sales View Creation Request for R/Sales View Creation Request.csv", show_col_types = FALSE) |>
    clean_names()

  # Process data
  df <- df |>
    filter_last_date() |>
    change_format() |>
    # select columns
    select(c(
      "plant",
      "sales_org",
      "material_number",
      "acc_assig_grp",
      "material_group1",
      "material_group2",
      "material_group3"
    ))

  # Write data
  write_csv(df, output_1, na = "")
  write_csv(df, output_2, na = "")
  print("Files are created")
}

main()

File: Sales_Billing/src/pandas/_create_sales_view.py
import pandas as pd
from pathlib import Path
from janitor import clean_names


# Path
path = Path(__file__).parent.parent.parent


# Functions
def filter_last_date(df):
    last_date = df["date"].max()  # find the last date
    df = df[df["date"] == last_date]  # filter the DataFrame
    return df


def add_acc_assig_grp(row):
    type_mapping = {
        "FERT": "01",
        "HAWA": "02",
        "HALB": "03",
        "ROH": "04",
    }
    return type_mapping.get(row["Type"], None)


def add_material_group(row):
    customer_mapping = {
        "HMC": "183",
        "KMC": "191",
        "KIA": "191",
        "GM": "105",
        "RSM": "193",
        "Ssangyong": "180",
        "SYM": "180",
        "Others": "199",
        "MOBIS": "317",
        "Mobis": "317",
        "MANDO": "384",
        "Mando": "384",
        "Webasto": "376",
        "VT": "360",
        "Conti": "394",
    }

    return customer_mapping.get(row["Customer"], "Unknown")


def change_format(df):
    df["plant"] = "0180"
    df["sales_org"] = "5668"
    df["acc_assig_grp"] = df.apply(add_acc_assig_grp, axis="columns")
    df["material_group1"] = df.apply(add_material_group, axis="columns")
    return df


def select_columns(df, selected_columns):
    df = df[selected_columns]
    return df


def main():

    # Filenames
    input_file = path / "data" / "Sales View Creation Request.csv"
    output_1 = "C:" / "LSMW" / "Sales_view.txt"
    output_2 = path / "output" / "Sales_view.txt"

    # Read data
    df = pd.read_csv(input_file).clean_names()

    # Process data
    df = (
        df.pipe(filter_last_date)
        .pipe(change_format)
        .pipe(
            select_columns,
            [
                "plant",
                "sales_org",
                "material_number",
                "acc_assig_grp",
                "material_group1",
                "material_group2",
                "material_group3",
            ],
        )
    )

    # Write data
    df.to_csv(output_1, index=False)
    df.to_csv(output_2, index=False)

    print("Files are created")


if __name__ == "__main__":
    main()
